### PR TITLE
Fix the RocksDB iteration logic for keys and entries.  The problems a…

### PR DESCRIPTION
**Pull request by jwlent55**
…ddressed include:

1) Key and entry iterators failed to strip prefix from keys (if required).
2) RocksdbStorageEngineTest failed to test BdbStorageEngine (all tests).
3) Both Key iterators would miss first entry and then iterate off the end.
With these changes I was able to remove the Ignore annotation from 2 tests.

### CODE REVIEW

**Comment by jwlent55** [comment]
Just starting to test with the RocksDB storage engine and first want to get the unit tests passing.  I don't think the single partition iteration logic is covered by the existing RocksDB unit tests, but, I did test that manually.

**Comment by FelixGV** [comment]
That's awesome @jwlent55!

I'll probably be able to review and merge this in later today.

Please keep us posted, here or on the mailing list, about your progress with RocksDB!

We're very interested to hear about how that goes (:

-F

File: test/unit/voldemort/store/rocksdb/RocksdbStorageEngineTest.java [hunk 1]
```diff
@@ -85,11 +83,9 @@ public void tearDown() throws Exception {
     @Override
     public void testTruncate() throws Exception {}
 
-    @Ignore
```
Comment (FelixGV) [inline test/unit/voldemort/store/rocksdb/RocksdbStorageEngineTest.java]: I think here you need to completely remove the override, otherwise, you are still overriding that particular test with an empty implementation.

**Comment by FelixGV** [comment]
Hi @jwlent55,

I don't understand what you mean by your second point:

> 2) RocksdbStorageEngineTest failed to test BdbStorageEngine (all tests).

Can you elaborate?

-F

**Comment by jwlent55** [comment]
Ooops.  Regarding the comment:

2) RocksdbStorageEngineTest failed to test BdbStorageEngine (all tests).

Cut'n'paste error (and vague).  Let me try again with more details

2) RocksdbStorageEngineTest failed to test RocksDbStorageEngine (all tests).  Instead both times the test suite runs I think it tests the subclass PartitionPrefixedRocksDbStorageEngine.

The current code defaults the rocksdbPrefixKeysWithPartitionId boolean to true:

VoldemortConfig.java
        this.rocksdbPrefixKeysWithPartitionId = props.getBoolean("rocksdb.prefix.keys.with.partitionid",
                                                                 true);

But in the unit test only overrides the value in the unit test suite if "true" was passed in:

RocksdbStorageEngineTest.java

if(this.prefixPartitionId) {
            voldemortConfig.setRocksdbPrefixKeysWithPartitionId(true);
        }

I just changed this to always set the boolean based on what is passed in.

**Comment by FelixGV** [comment]
Ah ok, sounds good!

Would you mind creating a new (squashed) commit with the two tests you fixed completely removed (rather than leaving the empty ```@Override``` function) and a commit message which references RocksDB rather than BDB?

I'll merge it in when it's ready.

Thanks a lot!

-F

File: test/unit/voldemort/store/rocksdb/RocksdbStorageEngineTest.java (hunk 1 above)
Comment (jwlent55) [inline test/unit/voldemort/store/rocksdb/RocksdbStorageEngineTest.java]: Agreed.  My mistake.  I did my testing using a version of the unit test code that did not include those ignore/overrides.  I only saw that code when I patched it over to git.  In a brain dead move I just removed the Ignore.

**Comment by jwlent55** [comment]
Will do and I may also roll in a small change to the RocksdbStorageEngineAPITest for the rocksdbPrefixKeysWithPartitionId boolean setting issue.

**Comment by jwlent55** [comment]
I have updated the pull request commit to address your code and commit text comments.  Thanks for the review. Let me know if you want more changes and/or I did not create the new squashed commit correctly.

**Comment by FelixGV** [comment]
Thanks a lot @jwlent55. LGTM (:

**Comment by FelixGV** [comment]
This is merged in. Thanks @jwlent55 (:
